pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java

```diff
@@ -459,7 +459,13 @@ private BrokerResponseNative handleSQLRequest(long requestId, String query, Json
     if (offlineBrokerRequest == null && realtimeBrokerRequest == null) {
       LOGGER.info("No server found for request {}: {}", requestId, query);
       _brokerMetrics.addMeteredTableValue(rawTableName, BrokerMeter.NO_SERVER_FOUND_EXCEPTIONS, 1);
-      return BrokerResponseNative.EMPTY_RESULT;
+      BrokerResponseNative brokerResponse = BrokerResponseNative.EMPTY_RESULT;
+      if (numUnavailableSegments > 0) {
```

Comment:
I think the intention of the PR https://github.com/apache/pinot/pull/7397 was to always add this exception to BrokerResponse if numUnavailableSegments is > 0 and this case was missed ? 

So may be we just do this once by moving the code at line 512 to proper place that covers all scenarios unconditionally ?

Reply:
It's a little tough to do with the way the code is structured. We can't return early when there's segments missing because many use cases still want partial results if possible. And there are several places in this function that short-circuit and return their own response before adding previous exceptions.

But I've moved the code up to create a list of exceptions we can add to. And I've added the missing segments exception to the timeout response as well.